tests/FakeItEasy.Analyzer.CSharp.Tests/ArgumentConstraintNullabilityMismatchAnalyzerTests.cs

```diff
@@ -0,0 +1,127 @@
+        {{
+            var foo = A.Fake<IFoo>();
+            A.CallTo(() => {0}).DoesNothing();
+        }}
+    }}
+
+    interface IFoo
+    {{
+        void NullableParam(int? x);
+        void OtherNullableParam(long? x);
+        void NonNullableParam(int x);
+        void OtherNonNullableParam(int x);
+    }}
+}}";
+
+        public static IEnumerable<object[]> SupportedConstraints =>
+            TestCases.FromObject(
+                "_",
+                "Ignored");
+
+        [Theory]
+        [MemberData(nameof(SupportedConstraints))]
+        public void Diagnostic_should_be_triggered_when_non_nullable_constraint_is_used_for_nullable_parameter(string constraint)
+        {
+            string completeConstraint = $"A<int>.{constraint}";
+            string call = $"foo.NullableParam({completeConstraint})";
+            string code = string.Format(CodeTemplate, call);
+
+            this.VerifyCSharpDiagnostic(
+                code,
+                new DiagnosticResult
+                {
+                    Id = "FakeItEasy0004",
+                    Message = $"Parameter 'x' of method 'NullableParam' is nullable, but argument constraint '{completeConstraint}' is not. Calls where this argument is null will not be matched.",
+                    Severity = DiagnosticSeverity.Warning,
+                    Locations = new[] { new DiagnosticResultLocation("Test0.cs", 9, 46) }
+                });
+        }
+
+        [Fact]
+        public void Diagnostic_should_not_be_triggered_for_That_constraint()
+        {
+            string completeConstraint = "A<int>.That.Matches(x => true)";
+            string call = $"foo.NullableParam({completeConstraint})";
+            string code = string.Format(CodeTemplate, call);
+
+            this.VerifyCSharpDiagnostic(code);
+        }
+
+        [Theory]
+        [MemberData(nameof(SupportedConstraints))]
+        public void Diagnostic_should_not_be_triggered_when_nullable_constraint_is_used_for_nullable_parameter(string constraint)
+        {
+            string completeConstraint = $"A<int?>.{constraint}";
+            string call = $"foo.NullableParam({completeConstraint})";
+            string code = string.Format(CodeTemplate, call);
+
+            this.VerifyCSharpDiagnostic(code);
+        }
+
+        [Theory]
+        [MemberData(nameof(SupportedConstraints))]
+        public void Diagnostic_should_not_be_triggered_when_non_nullable_constraint_is_used_for_non_nullable_parameter(string constraint)
+        {
+            string completeConstraint = $"A<int>.{constraint}";
+            string call = $"foo.NonNullableParam({completeConstraint})";
+            string code = string.Format(CodeTemplate, call);
+
+            this.VerifyCSharpDiagnostic(code);
+        }
+
+        [Theory]
+        [MemberData(nameof(SupportedConstraints))]
+        public void Diagnostic_should_not_be_triggered_when_non_nullable_constraint_is_used_for_other_nullable_parameter(string constraint)

```

Comment:
> It's never occurred to me to supply a `A<int>._` for a `long?` (or even long) parameter.
> Is there any reason not to warn? It won't match a null.

Good point. I wanted to avoid false positives, so I made the analyzer quite restrictive on when it was triggered. But I don't remember my exact reasoning when I did this... something to do with not having to deal with implicit conversions, I think. But maybe it should be triggered in that case as well.

> (On that note, neither would it match a non-null. Maybe we should have a diagnostic for the case where the type of the constraint is mismatched!)

Oh, I had never thought of that. Currently `A<int>._` won't match if the actual parameter type is `long`. Yes, we should totally add a diagnostic for this.

> (Oh, I guess that's a good reason not to trigger this diagnostic. Even aligning the nullability won't cause a match.)

Indeed!